### PR TITLE
YouTube album/playlist recognition & scrobbling

### LINKED ISSUE #122: Youtube albums and concerts scrobbling
On youtube it's common practice to upload whole album, or a live convert. It's very useful feature. But it can't be scrobbled by every song.

Often such videos have track list in description. Often it's easy to parse:
(Here are examples)

```
04. Track title - 08:15
...
16 - Track title (23:42)
...
or without track number
03:14 Track Title
```

Will be great tool if scrobbler will parse this list and scrobbles song if it's time for it.
(is it possible to get current position of video on youtube - I don't really know)

### PR DESCRIPTION
**Pull request by janbaykara**
Closes #122, #655

This PR adds recognition of youtube playlists, by searching for lists of timestamp'd songs in the description. As `currentTime` goes on, it will scrobble the relevant track. It also adds generic, complementary `Connector` functionality for parsing playlists if one is recognised.
#### Connector class changes:
- **New:** `Connector.setThrottleInterval()` - optionally throttles `Connector.stateChangedWorker()`
- **New:** `Connector.getPlaylist()` - looks for & returns a standardised playlist Object of tracks
- **New:** `Connector.playlistIncrementCheck()` - gets the current track, by comparing `Connector.getPlaylist()` against `Connector.getCurrentTime()`
- **Changed:** `Connector.stateChangedWorker()` - now scrobbles identified tracks from `Connector.getPlaylist()` via `Connector.playlistIncrementCheck()`, if a playlist is identified.
##### Should work with:
- https://youtube.com/watch?v=iyMZl8glVYw (gets `artist` from YouTube title, `track`s from playlist)
- https://youtube.com/watch?v=N-3qILKnSrM (HH:MM:SS timestamps)
- https://youtube.com/watch?v=EzjX0QE_l8U (compilation of different artistTrack)
- https://youtube.com/watch?v=YKkOxFoE5yo (enclosed timestamps + manual timestamp recognition)
- https://youtube.com/watch?v=VFOF47nalCY (funky HTML entities + false-positive artistTrack ID'ing)
- https://youtube.com/watch?v=RW3li2XWgoM (strange jumble of number, timestamp, track name)
- https://youtube.com/watch?v=i5mQ_m5zm6U (PRESERVE lots of numbers all over the place)
- https://youtube.com/watch?v=4Xg1usuLpsM (Gets playlist from a comment)
- https://www.youtube.com/watch?v=9-NFosnfd2c (Removes ridiculous other duration timestamps)
##### Should **not** work with:
- https://youtube.com/watch?v=EfcY9oFo1YQ (incorrect timestamps => invalid playlist)
- https://youtube.com/watch?v=YqeW9_5kURI (a normal, single-track video)

### CODE REVIEW

File: connectors/v2/youtube.js [hunk 1]
```diff
@@ -22,14 +22,14 @@ Connector.currentTimeSelector = '#player-api .ytp-time-current';
...
-	}
-};
+// Connector.getUniqueID = function() {
+// 	var url = window.location.href;
+// 	var regExp = /^.*((youtu.be\/)|(v\/)|(\/u\/\w\/)|(embed\/)|(watch\?))\??v?=?([^#\&\?]*).*/;
+// 	var match = url.match(regExp);
+// 	if (match && match[7].length==11){
+// 		return match[7];
+// 	}
+// };
```
Comment (janbaykara) [inline connectors/v2/youtube.js]: Should renable this, append `#N` to end of UID in `Connector.js`

File: connectors/v2/youtube.js [hunk 2]
```diff
@@ -51,12 +51,48 @@ Connector.getArtistTrack = function () {
...
+};
+
+Connector.getPlaylist = function() {
+	var playlist = [];
+	var $container = $('#eow-description');
+
+	// for each line
+	var potentialTracks = $container.html().split(/\r\n|\r|\n|<br>/g);
+
+	_.each(potentialTracks.reverse(), function(maybeTrack) {
```
Comment (janbaykara) [inline connectors/v2/youtube.js]: `.reverse()` here is useless

File: connectors/v2/youtube.js [hunk 3]
```diff
@@ -51,12 +51,48 @@ Connector.getArtistTrack = function () {
...
+			var entry = {};
+			entry.startTime = Connector.stringToSeconds($timestamp.first().text());
+			$timestamp.remove();
+			entry.track = cleanseTrack($maybeTrack.text());
+			playlist.push(entry);
+		}
+	});
+
+	if(playlist.length <= 1) { return; }
+
```
Comment (janbaykara) [inline connectors/v2/youtube.js]: Sort playlist by startTime low -> high:

`
_.sortBy(playlist, function(track) { return track.startTime; })
`

File: core/content/connector.js [hunk 4]
```diff
@@ -285,137 +328,169 @@ var BaseConnector = window.BaseConnector || function () {
...
+				if (separator !== null) {
+					if(newPlaylistTrack.artist) { newArtist = newArtistTrack.substr(0, separator.index); }
+					if(newPlaylistTrack.track)  { newTrack = newArtistTrack.substr(separator.index + separator.length); }
+				}
 
-			// take action if needed
-			if (changedFields.length > 0 && this.reactorCallback !== null) {
-				this.reactorCallback(currentState, changedFields);
+				var trackIndex = this.currentPlaylist.indexOf(newPlaylistTrack);
+				newDuration = newPlaylistTrack.startTime + (this.currentPlaylist[trackIndex+1].startTime || 20 );
```
Comment (janbaykara) [inline core/content/connector.js]: Give UID for playlist row, ontop of track UID.

`
if(newUID) {
   newUID += "#"+trackIndex
}
`

**Comment by fedorg** [comment]
https://www.youtube.com/watch?v=YKkOxFoE5yo
This just needs some more recognition patterns.
I'll probably fix this soon

**Comment by janbaykara** [comment]
@fedorg Good spot. We can additionally look for `/[0-9]{2}:[0-9]{2}/` (00:00) regex patterns.

**Comment by galeksandrp** [comment]
Note: yt not recoginze hh:mm:ss timestamps, but not mm:ss longer than 1 hour.

File: connectors/v2/youtube.js [hunk 5]
```diff
@@ -51,12 +51,48 @@ Connector.getArtistTrack = function () {
...
+	var $container = $('#eow-description');
+
+	// for each line
+	var potentialTracks = $container.html().split(/\r\n|\r|\n|<br>/g);
+
+	_.each(potentialTracks.reverse(), function(maybeTrack) {
+		var $maybeTrack = $('<div>'+maybeTrack+'</div>');
+		var $timestamp = $maybeTrack.find('a[href=\'#\'][onclick]');
+
+		if($timestamp.length) {
```
Comment (janbaykara) [inline connectors/v2/youtube.js]: `
var timestampRegex = /[0-9]{2}:[0-9]{2}/;
if( $timestamp.length || timestampRegex.test(maybeTrack) ) {
`

File: connectors/v2/youtube.js [hunk 6]
```diff
@@ -51,12 +51,48 @@ Connector.getArtistTrack = function () {
...
+			var entry = {};
+			entry.startTime = Connector.stringToSeconds($timestamp.first().text());
+			$timestamp.remove();
```
Comment (janbaykara) [inline connectors/v2/youtube.js]: Do more to remove timestamp delimiters.

* **E.g. 1:** `00:00 - Trackname` -> Should remove the ` - ` (and other similar delimiters)
* **E.g. 2:** `Trackname [00:00]` -> Should remove the `[` and `]` (and other similar delimiters)

**Comment by janbaykara** [comment]
It now looks for comment playlists too, now, and should be far more robust in its recognition of valid playlists.

File: connectors/v2/youtube.js [hunk 7]
```diff
@@ -51,12 +51,95 @@ Connector.getArtistTrack = function () {
...
+		}
+		if (potentialPlaylist.length > 1) {
+			playlist = potentialPlaylist;
+			found = true;
+		} else {
+			i++;
+		}
+	}
+
+	if(playlist.length <= 1) { return; }
```
Comment (janbaykara) [inline connectors/v2/youtube.js]: Open question here: should the threshold for 'playlist' be higher than 1? 
* Maybe it should be 3+ for comments, 2+ for descriptions?

**Comment by janbaykara** [comment]
[ ] Get artistTrack (e.g. https://www.youtube.com/watch?v=EzjX0QE_l8U) and artist+track parsing working.

**Comment by fedorg** [comment]
https://www.youtube.com/watch?v=i5mQ_m5zm6U
The first track is not recognized, 3rd track is missing a part of its name.

**Comment by janbaykara** [comment]
@fedorg Interesting. Perhaps I was a bit too draconian with the playlist numbering sanitsation.

**Comment by fedorg** [comment]
:+1:
Btw found an interesting one but this is kinda extreme:
https://www.youtube.com/watch?v=EfcY9oFo1YQ
Your stuff isnt fully compatible with afx I believe :)

**Comment by janbaykara** [comment]
@fedorg ... I'm... speechless. I suppose this would be one of those 'disable scrobbling' moments.

Actually, I suppose we could look to see if all the tracklisting times are within the duration of the video - that'd invalidate your AFX example where a timestamp of 5:26 is beyond the end time of 5:17.

**Comment by janbaykara** [comment]
So. Many. Edge-cases.

**Comment by janbaykara** [comment]
I've done a bunch of testing on countless Youtube album/compilation videos, ironing out edgecases, and I'm pretty happy with what we have now. Cheers @fedorg for your contributions - reckon this is good to go? If so we await your comments/suggestions @david-sabata :)

**Comment by fedorg** [comment]
Hell yeah, this works perfectly on my machine - 3 of them in fact, and the code is fairly robust IMO. Almost too good to be true :) Absolutely epic PR!

**Comment by janbaykara** [comment]
Something just occurred to me - OK, two things did:

1. **Improved comment playlist scanning via auto-scroll** - Youtube getPlaylist searches comments if it can't find one in the description, but Youtube only loads comments if you manually scroll down the page - we should figure out a way to force comment loading via the connector, to increase chances of playlist recognition.

2. **Convenient popup playlist navigation** - I was listening to a compilation just now, and wanted to navigate back to a previous track in the video, so I scrolled down through the comments to find the playlist timestamps. *But* what if that playlist, once found, was copied to the popup or something, for quick navigation?

E.g.
fromplaylist

**Comment by david-sabata** [comment]
Oh I don't like auto-scrolling :-P
YouTube is intentionally not loading comments when you don't need them and they know why they do it. If there's track list in comments, there's also chance that there are multiple comments with times, probably conflicting with each other... I don't have any numbers, but I think scanning description would be great move from where we are now. Let's try that and see. What do you think?
If this was to be implemented anyway, I'd leave it as non-default connector option.

Quick navigation looks cool :smiley: 
But it would be shame if this was only for YouTube. We could have some optional extension to connector API, that could work with the player. This functionality is quite far from _just scrobbling_, but as you said, it may be handy.

**Comment by janbaykara** [comment]
@david-sabata the quick nav would of course be backed by the site-agnostic `Connector.getPlaylist()` method. However, I'm also aware that it goes beyond the spirit of this PR, which is to get more granular about scrobbling songs. I'll have a mull over that and maybe submit a specific PR for playlist navigation in future.

In terms of comment scanning:
* **Ok,** I can agree that auto-scrolling is a naff idea; let's forget that for now (although... youtube comment autoloading could... idk, I may have another look at that in future)
* **But!** I am strongly in favour of comment-scanning, when it's possible. YouTube puts the most-upvoted comments at the top of the list, which means the most useful and valid playlist timestamps will rise to the top. 50% of the videos I've tested this against have made use of user-submitted playlists, and they've all been accurate as they are the most-upvoted ones.

The spirit of this extension is to make the internet scrobble-able. Surely recognising collections of songs on YouTube, and letting users scrobble those songs, fits firmly into that goal.

**Comment by fedorg** [comment]
I've just made a patch which downloads the track/album data from lastfm, please don't merge this PR just yet - my commits will be up in several hours after I've tested it enough.

**Comment by janbaykara** [comment]
Another one to test: https://www.youtube.com/watch?v=gTH7XefE85A

**Comment by david-sabata** [comment]
@fedorg did you want to add some more commit here?

@janbaykara sorry for the long delay. I've just merged a ton of PRs. Can you please rebase?

**Comment by fedorg** [comment]
Yes, but I can't modify this PR unless @janbaykara accepts this https://github.com/janbaykara/web-scrobbler/pull/8 , which I think he is reluctant to do.

**Comment by janbaykara** [comment]
@david-sabata @fedorg Bit busy, will have a look at this in the next week or two.

**Comment by janbaykara** [comment]
Hey @fedorg, I'm having a little trouble merging Master back in to Youtube PR - you added a whole bunch of funky mechanics that are making my head ache. Mind lending a hand?

**Comment by fedorg** [comment]
Sure thing, I'll try to simplify the code as I go.

**Comment by fedorg** [comment]
Ok, did the integration, haven't had time for cleanup yet, but it works just fine.

**Comment by janbaykara** [comment]
I barely recognise it anymore tbh, so cleanup is a bit beyond me. This connector could probably do with a bit of high-level documentation further down the line...

**Comment by fedorg** [comment]
I am working on cleanup right now, I wasn't implying that someone else should do it. It may take a while, though.

**Comment by david-sabata** [comment]
No problem. Clean code is preferred before quick delivery ;)

**Comment by janbaykara** [comment]
@fedorg I'm probably not going to have time to finish this PR up tbh, so feel free to fork and submit it.

**Comment by Miyou** [comment]
Been playing around with this PR and it has been working great :) However, I just found a video where the artist being scrobbled is wrong: https://www.youtube.com/watch?v=cW1TJ5dvmik
The songs are being scrobbled as Entheogenic instead of Ott.

**Comment by alexesprit** [comment]
Closed as outdated.

**Comment by zackkitzmiller** [comment]
@alexesprit Is there a new PR tracking this work or has the feature been abandoned.?

**Comment by killd0zr** [comment]
Feature is very needed. Scrobbling while listening mixes became useless

**Pull request closed without merging**